### ISSUE TITLE
Index Checker unable to infer length properties of CharSequence length() method

### ISSUE THREAD

**Issue by gokaco** [typetools/checker-framework#2561]
Consider the following code:-
```
class ex{
	public static void pr(CharSequence ch) {
		char[] newCh = new char[ch.length()];
		for (int i = 0; i < newCh.length; i++)
			newCh[i] = ch.charAt(i);
	}
}
```
Command:-
```
javac -processor index ex.java
```
Output:-
```
ex.java:7: error: [argument.type.incompatible] incompatible types in argument.
        newCh[i] = ch.charAt(i);
                             ^
  found   : @LTLengthOf(value="newCh", offset="0") int
  required: @LTLengthOf(value="ch", offset={}) int
1 error
```
If I change code to:-
```
class ex{
	public static void pr(CharSequence ch) {
		String b=ch.toString();
		char[] newCh = new char[b.length()];
		for (int i = 0; i < newCh.length; i++)
			newCh[i] = ch.charAt(i);
	}
}
```
then it works fine.

**Comment by kelloggm** [comment]
This is by design: a `String` is immutable (meaning its length cannot change), but a `CharSequence` could be a mutable-length collection (for instance, it might be a `StringBuilder`. The interaction between `CharSequence` and `String` in the Index Checker can be somewhat confusing because `String` is a subtype of `CharSequence`, but the Index Checker treats them differently. When designing the checker, we decided to accept that it was confusing so that the checker could soundly handle strings, which are important in real programs.

**Comment by gokaco** [comment]
@kelloggm This error occured in travis while building this pull request- #2523 . How to suppress this error then?
Link to travis- [link](https://travis-ci.org/typetools/checker-framework/jobs/547170907#L1925)

Thanks

**Comment by kelloggm** [comment]
You should suppress both warnings and include a link to this issue, since it contains an explanation of the design.

**Comment by gokaco** [comment]
@kelloggm This error is due to typetools/guava [Ascii.java](https://github.com/typetools/guava/blob/master/guava/src/com/google/common/base/Ascii.java) occurring at line number 438 and 488. Do you mean to say that these function [toLowerCase](https://github.com/typetools/guava/blob/0d3dafcb1e50702d0838f12c3313d19331bc1f2e/guava/src/com/google/common/base/Ascii.java#L432) and [toUpperCase](https://github.com/typetools/guava/blob/0d3dafcb1e50702d0838f12c3313d19331bc1f2e/guava/src/com/google/common/base/Ascii.java#L482) are not correctly annotated as I can see in the class that many of the methods have @SuppressWarnings on them but not these two methods ?
And they are performing same type of functions as you suggested Index Checker does not support.
Please check.

Thanks

**Comment by kelloggm** [comment]
@gokaco:

Please surround any annotation with backticks (\`); otherwise a user will get get pinged.

For you question: Yes, it looks like they were incorrectly annotated. I don't know why no warning was issued before - I would expect those to always issue an error. Perhaps the code has recently been updated? Regardless, the warnings should be suppressed, because they are false positives (assuming that there is no concurrent modification to the `CharSequence`, which is fine).

**Comment by mernst** [comment]
> When designing the checker, we decided to accept that it was confusing so that the checker could soundly handle strings

Why would handling `@SameLen` between a String and a CharSequence lead to unsoundness?

The Index Checker could refine the `@SameLen` types of the CharSequence and probably the String.  Then, at any possible side effect to the CharSequence, remove the refinement.
But maybe I missed something.  Can you explain what is fundamentally unsound about that approach?

**Comment by kelloggm** [comment]
> Then, at any possible side effect to the CharSequence, remove the refinement

I don't think there's anything that makes this fundamentally impossible, but it is inconvenient: the `CharSequence` could have aliases, so any call to a non-pure method would require us to undo the refinement.

There isn't a facility in the SameLen Checker (or, as far as I know, anywhere else in the Checker Framework) to remove just part of a refined SameLen annotation. So we would lose all information about both the `CharSequence` and the `String`. If I recall correctly, we made the decision that it would be preferable to:
* keep refinements to Strings when methods are called, and
* not open the Pandora's box of `@SameLen` annotations on mutable sequences, which we don't support in general.

**Comment by mernst** [comment]
Your message when you closed the issue said that the issue was soundness.
Now, you are saying it's inconvenient.  In that case, the issue should be re-opened.  Or maybe open a new one since this one now has a long, misleading preamble discussion.

Throughout the Checker Framework, a call to a side-effecting method undoes lots of refinements.  The Index Checker can do the same.  You are right that we would want to only remove the mutable expressions from the `@SameLen` annotation; I think that is manageable.

It may be that earlier we made a decision based on experience to date, but with more experience we can change that decision.

**Comment by kelloggm** [comment]
> Your message when you closed the issue said that the issue was soundness.

Yes. And it still is - we don't know whether it is sound or not, as I alluded to in my last message:

> not open the Pandora's box of @SameLen annotations on mutable sequences, which we don't support in general.

I'm saying that we don't know if doing so would be sound, and *in addition* it is inconvenient.
